README.md
# cough-speech-sneeze

Converts the [cough-speech-sneeze-raw-data] data set to the [Unified Format] using
[audata], and publishes it again on Artifactory.

## Prerequisites

[Java] must be installed.

## Publishing

Run `./gradlew publish`.

An [Artifactory] account with deploy permissions must be configured via Gradle properties `artifactoryUser` and `artifactoryApiKey`.


[cough-speech-sneeze-raw-data]: https://gitlab.audeering.com/data/cough-speech-sneeze-raw-data
[Unified Format]: http://tools.pp.audeering.com/audata/data-format.html
[audata]: https://gitlab.audeering.com/tools/pyaudata
[Artifactory]: http://artifactory.audeering.local/
[Java]: https://sdkman.io/sdks#java


CHANGELOG.md
Changelog
=========


Version 2.0.0 -- 2021-06-18
---------------------------

* Added: Coughing and sneezing files containing other sounds
  or of bad quality are deleted


Version 1.0.1 -- 2021-01-28
---------------------------

* Added: conversion to new audb format


Version 1.0.0 -- 2020-10-30
---------------------------

* Added: initial release


cough-speech-sneeze-metadata/python/test_filenames.py
import os

import pandas as pd

import audeer
import audiofile as af

import audata.define as define
import audata.utils as utils
from audata import (
    AudioInfo,
    Column,
    Database,
    Scheme,
    Table
)


data_root = 'build/unpacked/'


CATEGORY ={
    'coughing': 'coughing',
    'silence': 'silence',
    'sneezing': 'sneezing',
    'speech': 'speech'
}


files = list(utils.scan_files(
        root=data_root,
        recursive=True,
        pattern='*.wav',
    ))


def get_category(filename):
    category_id = audeer.basename_wo_ext(filename).split('-')[2]
    return CATEGORY[category_id]


category_files = [get_category(f) for f in files]


2.0.0/publish.py
import pandas as pd
import audeer
import audb


name = 'cough-speech-sneeze'
previous_version = '1.0.0'
version = '2.0.0'
build_dir = '../build'
repository = audb.Repository(
    name='data-public-local',
    host='https://artifactory.audeering.com/artifactory',
    backend='artifactory',
)

build_dir = audeer.mkdir(build_dir)

db = audb.load_to(
    build_dir,
    name,
    version=previous_version,
    num_workers=8,
)

#loading csv file containing reevaluated labels for cough and sneeze
path_csv_reevaluation = '../20210412-102437-cough-sneeze/20210412-102437_cough-and-sneeze_annotations-cough_sneeze.csv'
df = pd.read_csv(path_csv_reevaluation)
df = df[['Media File','Answer 1','Bad File?']]

#Removing files with bad sound quality or other sounds than cough and sneeze
df = df[df['Bad File?'] == False ]  
df_cough = df[df['Answer 1'] == 'coughing']
df_sneeze = df[df['Answer 1'] == 'sneezing']

#Creating new column that matches the file name in the original csv file
df_cough['file'] = 'coughing/' + df_cough['Media File'].astype(str)
df_sneeze['file'] = 'sneezing/' + df_sneeze['Media File'].astype(str)

#loading the original csv file for the data base
df_old = pd.read_csv(build_dir +'/db.files.csv')

#seperating csv file into sub catagories
df_old_cough = df_old[df_old['category'] == 'coughing']
df_old_sneeze = df_old[df_old['category'] == 'sneezing']

#Removing cough and sneeze files with were not accepted in 2nd scoring round
for x in df_old_cough['file'].copy():
    if not df_cough['file'].str.contains(x).any():
        db.drop_files(x)
        
for x in df_old_sneeze['file'].copy():
    if not df_sneeze['file'].str.contains(x).any():
        db.drop_files(x)

# Update header metadata
db.license = 'Unknown'

db.description = (
    'Cough-speech-sneeze: a data set of human sounds This dataset was collected by' 
    'Dr. Shahin Amiriparian. It contains samples of human speech, coughing, and' 
    'sneezing collected from YouTube, as well as silence clips. The original' 
    'publication of this (possibly then extended) dataset is the following:' 
    'Amiriparian, S., Pugachevskiy, S., Cummins, N., Hantke, S., Pohjalainen, J.,' 
    'Keren, G., Schuller, B., 2017. CAST a database: Rapid targeted large-scale big' 
    'data acquisition via small-world modelling of social media platforms, in: 2017' 
    'Seventh International Conference on Affective Computing and Intelligent' 
    'Interaction (ACII). IEEE, pp. 340–345. https://doi.org/10.1109/ACII.2017.8273622'
)
db.save(build_dir)

audb.publish(
    build_dir,
    version,
    repository,
    previous_version=previous_version,
)


version-2.0.0./cough_sneeze_update.py
import pandas as pd
import numpy as np
import os
import shutil 

#loading csv file containing reevaluated labels for cough and sneeze
path_csv = '/home/audeering.local/jthybo/Documents/sneeze_cough_data/20210412-102437-cough-sneeze/'
df = pd.read_csv(path_csv +'20210412-102437_cough-and-sneeze_annotations-cough_sneeze.csv')
df = df[['Media File','Answer 1','Bad File?']]


#Removing files with bad sound quality or other sounds than cough and sneeze
df = df[df['Bad File?'] == False ]  
df_cough = df[df['Answer 1'] == 'coughing']
df_sneeze = df[df['Answer 1'] == 'sneezing']

#Creating new column that matches the file name in the original csv file
df_cough['file'] = 'coughing/' + df_cough['Media File'].astype(str)
df_sneeze['file'] = 'sneezing/' + df_sneeze['Media File'].astype(str)

#loading the original csv file for the data base

path_old_csv = '/home/audeering.local/jthybo/audb/cough-speech-sneeze/1.0.0/d3b62a9b/'
df_old = pd.read_csv(path_old_csv +'db.files.csv')

#seperating csv file into sub catagories
df_old_cough = df_old[df_old['category'] == 'coughing']
df_old_sneeze = df_old[df_old['category'] == 'sneezing']
df_old_silence = df_old[df_old['category'] == 'silence']
df_old_speech = df_old[df_old['category'] == 'speech']


v=0
#Removing cough and sneeze files with were not accepted in 2nd scoring round
for x in df_old_cough['file'].copy():
    if df_cough['file'].str.contains(x).any():
     v=v+1
    else:
        df_old_cough = df_old_cough[~df_old_cough['file'].str.contains(x)]

for x in df_old_sneeze['file'].copy():
    if df_sneeze['file'].str.contains(x).any():
     v=v+1
    else:
        df_old_sneeze = df_old_sneeze[~df_old_sneeze['file'].str.contains(x)]


#Create foldes for version 2.0.0
dirName = 'audb/cough-speech-sneeze/2.0.0/d3b62a9b/'
eventName = ['coughing','silence','sneezing','speech']

for name in eventName:
    os.makedirs(dirName+name)

#Moving accepted files from old database to the new
for file in df_old_cough['file']:
    source = path_old_csv + 'coughing/' + file[9:]
    destination = dirName + 'coughing/'
    shutil.move(source, destination)

for file in df_old_sneeze['file']:
    source = path_old_csv + 'sneezing/' + file[9:]
    destination = dirName + 'sneezing/'
    shutil.move(source, destination)

for file in df_old_silence['file']:
    source = path_old_csv + 'silence/' + file[8:]
    destination = dirName + 'silence/'
    shutil.move(source, destination)

for file in df_old_speech['file']:
    source = path_old_csv + 'speech/' + file[6:]
    destination = dirName + 'speech/'
    shutil.move(source, destination)

#Saving update csv file
final_csv_file = pd.concat([df_old_cough,df_old_speech,df_old_silence,df_old_sneeze], ignore_index=True)
final_csv_file.to_csv('audb/cough-speech-sneeze/2.0.0/d3b62a9b/db.files.csv')



previous/cough-speech-sneeze-metadata/python/convert.py
import os

import pandas as pd

import audiofile as af

import audata.define as define
import audata.utils as utils
from audata import (
    AudioInfo,
    Column,
    Database,
    Scheme,
    Table
)


CATEGORY = [
    'coughing',
    'silence',
    'sneezing',
    'speech'
]


def convert(
        description: str,
        data_root: str,
        annotation_root: str = None
) -> Database:
    r"""Conversion function called by Gradle plugin."""

    ##########
    # Header #
    ##########
    db = Database(
        name='cough-speech-sneeze',
        source='Dataset based on the publication of Shahin Amiriparian: "Amiriparian, S., Pugachevskiy, S., Cummins, N., Hantke, S., Pohjalainen, J., Keren, G., Schuller, B., 2017. CAST a database: Rapid targeted large-scale big data acquisition via small-world modelling of social media platforms, in: 2017 Seventh International Conference on Affective Computing and Intelligent Interaction (ACII). IEEE, pp. 340–345. https://doi.org/10.1109/ACII.2017.8273622"',
        usage=define.Usage.COMMERCIAL, # CC-4
        description=description,
    )

    #########
    # Media #
    #########
    # used `soxi <file name> to sample some (~10) audio files from all categories and checked their metadata
    db.media['microphone'] = AudioInfo(
        sampling_rate=None, # all categories got 44100 kHz, but 'silent' only got 16000 kHz
        channels=1,
        format='wav',
    )

    ###########
    # Schemes #
    ###########
    db.schemes['category'] = Scheme(labels=CATEGORY)
    db.schemes['duration'] = Scheme(dtype=define.DataType.TIME)

    ##########
    # Tables #
    ##########
    files = list(utils.scan_files(
        root=data_root,
        recursive=True,
        pattern='*.wav',
    ))
    category_files = [get_category(f) for f in files]
    durations = utils.run_worker_threads(
        num_workers=12,
        task_fun=lambda x: pd.to_timedelta(af.duration(
            os.path.join(data_root, x)), unit='s'),
        params=files,
        progress_bar=True,
        task_description='Parse duration',
    )
    # files
    db['files'] = Table(files=files)
    db['files']['category'] = Column(scheme_id='category')
    db['files']['category'].set(category_files, files=files)
    db['files']['duration'] = Column(scheme_id='duration')
    db['files']['duration'].set(durations, files=files)

    return db


def get_category(filename):
    # The category-folder is always the 2nd last part of the string (e.g. 'coughing/ipo39x2bv9c_12.7253-13.7358.wav')
    return filename.split('/')[-2]
